src/main/kotlin/com/dprint/formatter/DprintExternalFormatter.kt

```diff
@@ -55,29 +57,77 @@ class DprintExternalFormatter : AsyncDocumentFormattingService() {
             return null
         }
 
+        if (!editorService.canRangeFormat() && isRangeFormat(formattingRequest)) {
+            return null
+        }
+
         return object : FormattingTask {
+            private var formattingId: Int? = null
+            private var isCancelled = false
+
             override fun run() {
                 val content = formattingRequest.documentText
+                val ranges = formattingRequest.formattingRanges
+                val future = CompletableFuture<FormatResult>()
+
+                for (range in ranges.subList(1, ranges.size)) {
```

Comment:
So apparently IntelliJ can specify multiple ranges in a single format. Atm this assumes that the output of each individual range format will be the input for the next. We need to test this when the time comes for range formatting but I don't think it happens often enough to support it in dprint.

Reply:
That's interesting. I'm wondering if we should instead change range formatting to be a list of ranges, but that would add a lot of complexity and doing this kind of pipeline approach doesn't seem too bad.